File: api/services/classifier/scorer.py
"""
api/services/classifier/scorer.py

Combines grammar features and vocabulary results into a final CEFR classification
with a confidence score.

Confidence starts at 1.0 and is reduced by:
  - Low-confidence grammar features
  - Unknown content words
  - No grammar features detected (ambiguous)
  - Many conflicting levels
"""

CEFR_ORDER = {"A1": 1, "A2": 2, "B1": 3, "B2": 4, "C1": 5, "C2": 6}


def _max_level(levels: list) -> str:
    """Return the highest CEFR level from a list, defaulting to A1."""
    if not levels:
        return "A1"
    return max(levels, key=lambda l: CEFR_ORDER.get(l, 0))


def calculate_result(phrase: str, grammar_features: list, vocab_result: dict) -> dict:
    """
    Compute the final classification result.

    Args:
        phrase:           Original input phrase.
        grammar_features: List of GrammarFeature from rules_de.
        vocab_result:     Dict from vocabulary.detect_vocabulary_level.

    Returns:
        Dict with keys: level, grammar_level, vocabulary_level,
                        confidence, grammar_features, vocabulary_features,
                        source, ai_used.
    """
    # ── Grammar level ────────────────────────────────────────────────────
    grammar_levels = [f.level for f in grammar_features]
    grammar_level  = _max_level(grammar_levels) if grammar_levels else "A1"

    # ── Vocabulary level ─────────────────────────────────────────────────
    vocabulary_level = vocab_result.get("level", "A1")

    # ── Overall = max of both ────────────────────────────────────────────
    overall_level = _max_level([grammar_level, vocabulary_level])

    # ── Confidence ───────────────────────────────────────────────────────
    confidence = 1.0

    if grammar_features:
        # Average confidence of detected features
        avg_feat_conf = sum(f.confidence for f in grammar_features) / len(grammar_features)
        # Pull overall confidence toward the feature confidence
        confidence = 0.7 * avg_feat_conf + 0.3 * confidence
    else:
        # No grammar features: confidence drops — only vocabulary or nothing found
        known_words = vocab_result.get("words", [])
        if known_words:
            # We have vocabulary evidence but no grammar features
            confidence -= 0.15
        else:
            # Completely blank — simple A1 sentence with no markers at all
            # Still likely A1, but we're not 100% sure
            confidence -= 0.20

    # Penalty for unknown content words
    unknown = vocab_result.get("unknown_count", 0)
    if unknown >= 4:
        confidence -= 0.25
    elif unknown >= 2:
        confidence -= 0.15
    elif unknown >= 1:
        confidence -= 0.05

    # Penalty if multiple very different levels detected (unlikely, but possible)
    unique_levels = set(grammar_levels)
    if len(unique_levels) >= 3:
        confidence -= 0.10

    confidence = round(max(0.0, min(1.0, confidence)), 3)

    # ── Reason generation ────────────────────────────────────────────────
    g_reasons = [f.name for f in grammar_features if f.level == overall_level or f.level == grammar_level]
    v_matching = [w["word"] for w in vocab_result.get("words", []) if w.get("level") == overall_level]

    reasons = []
    if g_reasons:
        reasons.extend(g_reasons)
    if v_matching:
        reasons.append(f"Словарь: {', '.join(v_matching)}")

    if not reasons:
        reason = "Präsens (базовая фраза)" if overall_level == "A1" else f"Уровень {overall_level}"
        reason_short = "Präsens" if overall_level == "A1" else overall_level
    else:
        # Deduplicate while preserving order
        seen = set()
        dedup = [r for r in reasons if not (r in seen or seen.add(r))]
        reason = " | ".join(dedup)
        # Short primary reason (first grammar rule or vocab)
        primary = g_reasons[0] if g_reasons else (f"Словарь ({v_matching[0]})" if v_matching else overall_level)
        # Simplify common verbose names for short badge display
        short_map = {
            "Perfekt (haben + Part.II)": "Perfekt (haben)",
            "Perfekt (sein + Part.II)": "Perfekt (sein)",
            "Passiv Präsens (wird + Part.II)": "Passiv (wird)",
            "Passiv Präteritum (wurde + Part.II)": "Passiv (wurde)",
            "Passiv Perfekt (…worden)": "Passiv Perfekt",
            "um…zu Konstruktion": "um…zu",
            "ohne…zu Konstruktion": "ohne…zu",
            "statt…zu Konstruktion": "statt…zu",
            "je…desto Konstruktion": "je…desto",
            "sein + zu + Infinitiv": "sein + zu",
            "sich lassen + Infinitiv": "sich lassen",
        }
        reason_short = short_map.get(primary, primary)

    return {
        "level":               overall_level,
        "reason":              reason,
        "reason_short":        reason_short,
        "grammar_level":       grammar_level,
        "vocabulary_level":    vocabulary_level,
        "confidence":          confidence,
        "grammar_features":    [
            {"name": f.name, "level": f.level, "confidence": f.confidence}
            for f in grammar_features
        ],
        "vocabulary_features": vocab_result.get("words", []),
        "source":              "rules",
        "ai_used":             False,
    }

File: app/src/services/classifier/scorer.js
/**
 * app/src/services/classifier/scorer.js
 *
 * Combines grammar features and vocabulary results into a final CEFR classification result.
 */

const CEFR_ORDER = { A1: 1, A2: 2, B1: 3, B2: 4, C1: 5, C2: 6 };

function maxLevel(levels) {
  if (!levels || levels.length === 0) return 'A1';
  return levels.reduce((max, lvl) => {
    return (CEFR_ORDER[lvl] || 0) > (CEFR_ORDER[max] || 0) ? lvl : max;
  }, 'A1');
}

export function calculateResult(phrase, grammarFeatures, vocabResult) {
  const grammarLevels = grammarFeatures.map(f => f.level);
  const grammarLevel  = grammarLevels.length > 0 ? maxLevel(grammarLevels) : 'A1';
  const vocabularyLevel = vocabResult.level || 'A1';
  const overallLevel = maxLevel([grammarLevel, vocabularyLevel]);

  let confidence = 1.0;

  if (grammarFeatures.length > 0) {
    const avgFeatConf = grammarFeatures.reduce((sum, f) => sum + f.confidence, 0) / grammarFeatures.length;
    confidence = 0.7 * avgFeatConf + 0.3 * confidence;
  } else {
    const knownWords = vocabResult.words || [];
    if (knownWords.length > 0) {
      confidence -= 0.15;
    } else {
      confidence -= 0.20;
    }
  }

  const unknown = vocabResult.unknown_count || 0;
  if (unknown >= 4) {
    confidence -= 0.25;
  } else if (unknown >= 2) {
    confidence -= 0.15;
  } else if (unknown >= 1) {
    confidence -= 0.05;
  }

  const uniqueLevels = new Set(grammarLevels);
  if (uniqueLevels.size >= 3) {
    confidence -= 0.10;
  }

  confidence = Math.round(Math.max(0.0, Math.min(1.0, confidence)) * 1000) / 1000;

  // ── Reason generation ────────────────────────────────────────────────
  const gReasons = grammarFeatures.filter(f => f.level === overallLevel || f.level === grammarLevel).map(f => f.name);
  const vMatching = (vocabResult.words || []).filter(w => w.level === overallLevel).map(w => w.word);

  const reasons = [];
  if (gReasons.length > 0) reasons.push(...gReasons);
  if (vMatching.length > 0) reasons.push(`Словарь: ${vMatching.join(', ')}`);

  let reason = '';
  let reasonShort = '';

  if (reasons.length === 0) {
    reason = overallLevel === 'A1' ? 'Präsens (базовая фраза)' : `Уровень ${overallLevel}`;
    reasonShort = overallLevel === 'A1' ? 'Präsens' : overallLevel;
  } else {
    const dedup = Array.from(new Set(reasons));
    reason = dedup.join(' | ');

    const primary = gReasons.length > 0 ? gReasons[0] : (vMatching.length > 0 ? `Словарь (${vMatching[0]})` : overallLevel);
    const shortMap = {
      'Modalsatz (so, wie)': 'so, wie',
      'Perfekt (haben + Part.II)': 'Perfekt (haben)',
      'Perfekt (sein + Part.II)': 'Perfekt (sein)',
      'Passiv Präsens (wird + Part.II)': 'Passiv (wird)',
      'Passiv Präteritum (wurde + Part.II)': 'Passiv (wurde)',
      'Passiv Perfekt (…worden)': 'Passiv Perfekt',
      'um…zu Konstruktion': 'um…zu',
      'ohne…zu Konstruktion': 'ohne…zu',
      'statt…zu Konstruktion': 'statt…zu',
      'Adjektiv + zu + Infinitiv': 'Adjektiv + zu',
      'je…desto Konstruktion': 'je…desto',
      'sein + zu + Infinitiv': 'sein + zu',
      'sich lassen + Infinitiv': 'sich lassen',
    };
    reasonShort = shortMap[primary] || primary;
  }

  return {
    level: overallLevel,
    reason,
    reason_short: reasonShort,
    grammar_level: grammarLevel,
    vocabulary_level: vocabularyLevel,
    confidence,
    grammar_features: grammarFeatures,
    vocabulary_features: vocabResult.words || [],
    source: 'rules',
    ai_used: false,
  };
}

